include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge.h

```diff
@@ -130,6 +130,25 @@ __find_start_point(const _Rng1& __rng1, const _Index __rng1_from, _Index __rng1_
     return _split_point_t<_Index>{*__res, __index_sum - *__res + 1};
 }
 
+template <typename _Rng1, typename _Rng2, typename _value_t_rng1 = oneapi::dpl::__internal::__value_t<_Rng1>,
+          typename _value_t_rng2 = oneapi::dpl::__internal::__value_t<_Rng2>>
+constexpr auto
+__can_use_ternary_op(int) -> decltype(true ? std::declval<_value_t_rng1>() : std::declval<_value_t_rng2>(),
+                                      std::true_type{})
+{
+    return {};
+}
+
+template <typename _Rng1, typename _Rng2>
+constexpr auto
+__can_use_ternary_op(long) -> std::false_type
+{
+    return {};
+}
+
+template <typename _Rng1, typename _Rng2>
+constexpr static bool __can_use_ternary_op_v = decltype(__can_use_ternary_op<_Rng1, _Rng2>(int{0}))::value;
```

Comment:
Why `int{0}` instead of `0`?

Reply:
*I think* normally with this pattern you don't need to worry about what type you pass in because the `...` overload will be a lower priority than all other options, but with the variadic function limitation we make the "default" another typed parameter instead.  If we were to pass in a `long` then it just will take the "default" even if the `int` overload is available, so we have to make sure that the optionally enabled overload has priority.

Thinking more about this, what about systems where `int` is defined the same as `long`?  Should we use more explicitly different types here?

Reply:
I agree `int{0}` is the same as `0`, but the `int{0}` provides more intention in the code.

Reply:
I think "int" is more preferable in overloads choosing context. We have many places in the code where we rely on that.
For example, https://github.com/uxlfoundation/oneDPL/blob/main/include/oneapi/dpl/pstl/utils_ranges.h#L47

Reply:
Perhaps we should use `std::int32_t` and std::int16_t`, and make sure our argument aligns with the type in the "true" overload, just to guarantee they are distinct types.

Reply:
> but the `int{0}` provides more intention in the code.

which "intention" do we have?

Reply:
I hope the same like you wrote here:
```C++
    //SFINAE iterator type checks
    template <typename It>
    static constexpr auto
    __is_addressable(int) -> decltype(std::addressof(*::std::declval<It&>()), std::true_type{});
    template <typename It>
    static constexpr std::false_type
    __is_addressable(...);
    template <typename It>
    static constexpr bool __is_addressable_v = decltype(__is_addressable<It>(0))::value;
```

Reply:
> which "intention" do we have?

Exactly this:
> make sure our argument aligns with the type in the "true" overload

If the type aligns with the default, it will just use the default no matter the validity of the SFINAE check in the "true" overload.  If the type doesn't align with either, then I think it is ambiguous, and will end up as an error.

Reply:
You have also passed `0` as argument.

Reply:
Let's finalize: my proposal - ask colleagues to check their compile error in our code after merge.
So right now - merge as is.

Reply:
https://godbolt.org/z/1ae3s3T8o

Hopefully this shows why it changes when we have `long` as the "default" overload as compared to `...`.

Switch between the alternatives there.  It makes it more important what type we pass in.  As I understand it `int`, `long`, etc.  can be defined by the implementation and may sometimes be exactly the same type.  
There is nothing magic about `int` and `long` here, so I believe we should instead pick more explicitly defined and distinct types.

Reply:
One more variant from @dmitriy-sobolev : https://godbolt.org/z/1GP5Y68cM

Reply:
The PR https://github.com/uxlfoundation/oneDPL/pull/2024 has been merged into this branch.

Reply:
@MikeDvorskiy is this conversation resolved now?